### PR TITLE
8268363: AArch64: Implement string_indexof_char intrinsic in SVE

### PR DESCRIPTION
**Pull request by tatwaichong**
This patch implements string_indexof_char intrinsic in SVE for both
UTF-16 and Latin-8 variants. In the current underneath implementation,
there is only 4 and 8 characters are compared at a time in UTF-16 and
Latin-8 respectively. In this patch, we adopt SVE to vectorize this
intrinsic so that a wider comparison can be performed each time.

I've run [1] with this patch on an SVE machine, and the performance of
StringIndexOfChar.latin1_mixed_char (length=65) and
StringIndexOfChar.utf16_mixed_char (length=65) are improved significantly,
while no visible regression for other cases with shorter strings.
(length=7, 8, 15, 16 and 32)

[1] https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268363](https://bugs.openjdk.java.net/browse/JDK-8268363): AArch64: Implement string_indexof_char intrinsic in SVE

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4567/head:pull/4567` \
`$ git checkout pull/4567`

Update a local copy of the PR: \
`$ git checkout pull/4567` \
`$ git pull https://git.openjdk.java.net/jdk pull/4567/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4567`

View PR using the GUI difftool: \
`$ git pr show -t 4567`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4567.diff">https://git.openjdk.java.net/jdk/pull/4567.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -540,6 +540,75 @@ void C2_MacroAssembler::string_indexof_char(Register str1, Register cnt1,
...
+    // Undo the index increment.
+    sub(idx, idx, vec_len);
+
+    // Crop the vector to find its location.
+    sve_brka(tmp_pdn, tmp_pg, tmp_pdn, false /* isMerge */);
+    add(result, idx, -1);
+    sve_incp(result, T, tmp_pdn);
+  BIND(DONE);
+}
+
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:611]: That really is nice! I do like SVE.

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 2]
```diff
@@ -16645,21 +16654,57 @@ instruct string_indexof_char(iRegP_R1 str1, iRegI_R2 cnt1, iRegI_R3 ch,
...
+  match(Set result (StrIndexOfChar (Binary str1 cnt1) ch));
+  predicate((UseSVE > 0) && (((StrIndexOfCharNode*)n)->encoding() == StrIntrinsicNode::L));
+  effect(TEMP ztmp1, TEMP ztmp2, TEMP pgtmp, TEMP ptmp, KILL cr);
+
+  format %{ "StringLatin1 IndexOf char[] $str1,$cnt1,$ch -> $result # use sve" %}
+
+  ins_encode %{
+    __ string_indexof_char_sve($str1$$Register, $cnt1$$Register, $ch$$Register, $result$$Register,
+                               as_FloatRegister($ztmp1$$reg), as_FloatRegister($ztmp2$$reg),
+                               as_PRegister($pgtmp$$reg), as_PRegister($ptmp$$reg), true /* isL */);
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64.ad:16707]: Please consider moving the SVE-specific pattern into aarch64_sve_ad.m4 and regenerating aarch64_sve.ad

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 3]
```diff
@@ -3245,6 +3245,64 @@ void mvnw(Register Rd, Register Rm,
     f(pattern, 9, 5), f(0b0, 4), prf(pd, 0);
   }
 
+// SVE integer compare vector (wide-element variants are not included)
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3248]: The correct title for this section is "SVE Integer comparisons", as in the Arm ® Architecture Reference Manual Supplement, The Scalable Vector Extension". Please make sure all of these sections in the assembler have the same names as those in the manual.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 4]
```diff
@@ -3245,6 +3245,64 @@ void mvnw(Register Rd, Register Rm,
...
+  }
+
+  INSN(sve_cmpeq, 0b1010);
+  INSN(sve_cmpne, 0b1011);
+  INSN(sve_cmpge, 0b1000);
+  INSN(sve_cmpgt, 0b1001);
+#undef INSN
+
+// SVE 64-bit while[cond] (32-bit variants are not included)
+#define INSN(NAME, decode)                                                \
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3264]: And here it should be "Predicate counted loop (SVE)".

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp [hunk 5]
```diff
@@ -3245,6 +3245,64 @@ void mvnw(Register Rd, Register Rm,
...
+#undef INSN
+
+  void sve_brka(PRegister pd, PRegister pg, PRegister pn, bool isMerge) {
+    starti;
+    f(0b00100101, 31, 24), f(0b00, 23, 22), f(0b01000001, 21, 14),
+    prf(pg, 10), f(0b0, 9), prf(pn, 5), f(isMerge ? 1 : 0, 4), prf(pd, 0);
+  }
+
+// SVE cnt[b|h|w|d]
+#define INSN(NAME, TYPE)                                                             \
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3286]: "Element count and increment scalar (SVE)"

**Comment by tatwaichong** [comment]
Address review comments.
I'm not sure I do the right review procedure as I git force-pushed the new patch to my forked repo.

File: src/hotspot/cpu/aarch64/aarch64_sve_ad.m4 [hunk 6]
```diff
@@ -919,3 +919,39 @@ instruct vmaskcast(vReg dst) %{
...
+  format %{ "StringLatin1 IndexOf char[] $str1,$cnt1,$ch -> $result # use sve" %}
+
+  ins_encode %{
+    __ string_indexof_char_sve($str1$$Register, $cnt1$$Register, $ch$$Register, $result$$Register,
+                               as_FloatRegister($ztmp1$$reg), as_FloatRegister($ztmp2$$reg),
+                               as_PRegister($pgtmp$$reg), as_PRegister($ptmp$$reg), true /* isL */);
+  %}
+  ins_pipe(pipe_class_memory);
+%}
+
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_sve_ad.m4:957]: You shouldn't write this stuff out twice: this is an m4 file, so please use the macro processor. There are plenty of examples.

**Review by theRealAph — APPROVED** [review]
Thanks, nice work.

**Review by nsjian — APPROVED** [review]

**Pull request closed without merging**